broadinstitute/rawls
# [CA-1781] Sync Billing Account Changes with Google

DO NOT MERGE UNTIL https://broadworkbench.atlassian.net/browse/CA-1844 IS COMPLETE
Running CA-1844 after this is merged will potentially corrupt the state of billing accounts on workspaces and billing projects.

RR: https://broadworkbench.atlassian.net/browse/CA-1781
This change extends the `WorkspaceBillingAccountActor` (nee `WorkspaceBillingAccountMonitor`) code to update billing account changes on v1 billing projects even if there are no workspaces in the billing project.

Using the new `BillingAccountChanges` table, the actor
- gets the *latest* billing account change to all billing projects, ignoring the previous unsync'ed changed
- attempts to update the billing account on the billing project google project, if it exists
- gets all workspaces in the billiong project
- attempts to update the billing account on the workspace google project, if v2, and writes the state of the change to the workspace (ie billing account and billing account error message)

I've changed this file to use the more modern akka-actor-typed interface. I believe this has helped tremedously with testing behaviour directly, removing the need for polling.

In a FiaB, manually verified:
- changing billing accounts for empty v1 billing projects works
- billing accounts of v2 workspace google projects are updated.


  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
